aws-lc: bump from 2.0.17 to 3.0.0

ytsssun opened the pull request:
<!--
Tips:
- Please read the CONTRIBUTING document to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->


**Description of changes:**
Bump `aws-lc` 2.0.17 to `3.0.0`


**Testing done:**
Tested building custom FIPS and non-FIPS AMI with this SDK change. 
I have a setup where I use an nginx pod to enforce cipher suite that is not supported by FIPS
```
# Not supported by FIPS
ssl_ciphers 'ECDHE-RSA-CHACHA20-POLY1305';
```

And use ctr images pull to verify the tls behavior.

For FIPS variant, the tls handshake failed as expected:
```
bash-5.1# ctr images pull localhost:8043/test:latest
INFO[0000] host will try HTTPS first since it is configured for HTTP with a TLS configuration, consider changing host to HTTPS or removing unused TLS configuration  host="localhost:8043"
INFO[0000] trying next host                              error="failed to do request: Head \"https://localhost:8043/v2/test/manifests/latest\": remote error: tls: handshake failure" host="localhost:8043"
ctr: failed to resolve reference "localhost:8043/test:latest": failed to do request: Head "https://localhost:8043/v2/test/manifests/latest": remote error: tls: handshake failure
```

For non-FIPS variant, the tls handshake succeeded and the command is returning the normal image not found response:
```
ctr images pull localhost:8043/test:latest
INFO[0000] host will try HTTPS first since it is configured for HTTP with a TLS configuration, consider changing host to HTTPS or removing unused TLS configuration  host="localhost:8043"
INFO[0000] trying next host - response was http.StatusNotFound  host="localhost:8043"
ctr: failed to resolve reference "localhost:8043/test:latest": localhost:8043/test:latest: not found
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on hashes/aws-lc:
```diff
@@ -1,2 +1,2 @@
-# https://github.com/aws/aws-lc/archive/refs/tags/AWS-LC-FIPS-2.0.17.tar.gz
-SHA512 (AWS-LC-FIPS-2.0.17.tar.gz) = acb61f71d4039112ab027db56a9b5b1f9b272d41bd014cb733a71f365db13a2ade4ecf86d988900b2f7c8429086caf379a59803bae3c7a00111da0b72b357a2a
+# https://github.com/aws/aws-lc/archive/AWS-LC-FIPS-3.0.0/aws-lc-AWS-LC-FIPS-3.0.0.tar.gz
```
Please add the patches per [libcrypto.spec](https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/2b861f06738a8584537b7c24900260d07cbf747c/packages/libcrypto/libcrypto.spec#L10), and document the steps:
```suggestion
# https://github.com/aws/aws-lc/archive/AWS-LC-FIPS-3.0.0/aws-lc-AWS-LC-FIPS-3.0.0.tar.gz
### Additional patches from AWS-LC-FIPS 3.0 branch:
### git clone https://github.com/aws/aws-lc
### cd aws-lc
### git checkout origin/fips-2024-09-27
### git format-patch --no-numbered AWS-LC-FIPS-3.0.0..
```

These steps could also go in `patches/aws-lc/README.md` alongside the patches, if you prefer.

ytsssun replied:
Done. Now building the sdk to test.

ytsssun replied:
@bcressey @arnaldo2792 

Looks like there are more patches coming from that branch as of last week.

https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/develop/packages/libcrypto/libcrypto.spec#L10-L19 We only had three of them when we patched the change in core-kit.
```
0001-Cherry-pick-BORINGSSL_bcm_text_hash-Go-utility-2221.patch
0002-Cherry-pick-Fix-out-of-bound-OOB-input-read-in-AES-X.patch
0003-Move-OCSP-ASN1-type-functions-to-public-header-2239.patch
0004-Add-test-around-OpenSSL-behavior-for-BIO_get_mem_dat.patch
0005-Cherry-pick-support-for-CMake-4.0-to-fips-2024-09-27.patch
0006-Remove-some-indirection-in-SSL_certs_clear.patch
0007-Add-SSL_CTX_use_cert_and_key-2163.patch
0008-fips-2024-09-27-cherry-pick-FIPS-Integrity-Hash-Tool.patch
```

ytsssun commented:
The lc build took a large chunk of memory, it caused OOM on my build instances. Will attempt the build again with this fix https://github.com/bottlerocket-os/bottlerocket-sdk/pull/270.

ytsssun commented:
Rebased from the tip of develop and pulled more patches since last time.

ytsssun commented on patches/aws-lc/0009-Adding-detection-of-out-of-bound-pre-bound-memory-re.patch:
````diff
@@ -0,0 +1,504 @@
+
+The efficacy of the added test was shown by changing the decrypt
+cipher-stealing code for example to:
+# ```diff
+# --- a/crypto/fipsmodule/aes/asm/aesni-xts-avx512.pl
+# +++ b/crypto/fipsmodule/aes/asm/aesni-xts-avx512.pl
+# @@ -2569,7 +2569,7 @@ ___
+#    vpshufb       %xmm10,%xmm8,%xmm8
+
+# -  vmovdqu       -0x10($input,$length,1),%xmm3
+# +  vmovdqu       -0x12($input,$length,1),%xmm3
+#    vmovdqu       %xmm8,-0x10($output,$length,1)
+#  ```
````
I modified this diff block with comment prefix (`#`) because it is a diff block in the commit message, not actual diff block to be applied. It would confuse `git` if not make them as comments.

ytsssun commented on patches/aws-lc/0009-Adding-detection-of-out-of-bound-pre-bound-memory-re.patch:
````diff
@@ -0,0 +1,504 @@
+- Other larger changes, e.g. -0x20, will have the same result.
+
+Another test changes the location of the written output
+# ```@@ -2607,7 +2607,7 @@ ___
+
+#    .L_done_${rndsuffix}:
+#    # store last ciphertext value
+# -  vmovdqu       %xmm8,-0x10($output)
+# +  vmovdqu       %xmm8,-0x11($output)
+#  ___
+#    }
+#  ```
````
I modified this diff block with comment prefix (`#`) because it is a diff block in the commit message, not actual diff block to be applied. It would confuse `git` if not make them as comments.

ytsssun commented:
Fixed patch [0009-Adding-detection-of-out-of-bound-pre-bound-memory-re.patch](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/266/files#diff-b16ec04bb3d62c99a2be92a89cc0255e9a8ac6855766a8d5bc13d3c3bd56a3c7) which would fail git when applying the patches.

bcressey commented on patches/aws-lc/0001-Avoid-weak-symbols-when-building-for-Go.patch:
```diff
@@ -1,34 +0,0 @@
-From 0fb9b17aa5d0d63d17e6914f7c34358db69e5647 Mon Sep 17 00:00:00 2001
-From: Ben Cressey <bcressey@amazon.com>
-Date: Thu, 4 Jan 2024 18:07:33 +0000
-Subject: [PATCH] Avoid weak symbols when building for Go.
```
This patch may need to come back.

ytsssun replied:
Adding this back. Will build a new custom sdk image to test the change.

bcressey reviewed (commented):
Can you also add the recent "Avoid mixing SSE and AVX in XTS-mode AVX512 implementation" patch?

ytsssun committed:
aws-lc: bump from 2.0.17 to 3.0.0

- Updated naming of the package src.
- Drop old patch 0002-Fix-definition-of-ScopedFile-in-tests.patch
for 2.0.17.
- Add patches for 3.0.0 from the AWS-LC-FIPS 3.0 branch.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun commented:
> Can you also add the recent "Avoid mixing SSE and AVX in XTS-mode AVX512 implementation" patch?

This patch is already included yes.

ytsssun commented on hashes/aws-lc:
```diff
@@ -1,2 +1,8 @@
+### git clone https://github.com/aws/aws-lc
+### cd aws-lc
+### git checkout origin/fips-2024-09-27
+### git format-patch --start-number=1001 AWS-LC-FIPS-3.0.0..
```
I gave it a new start number here to avoid prefix collision with our first party patches not coming from aws-lc fips branch.

KCSesh commented:
Just want to see the testing output before approving!

bcressey commented on hashes/aws-lc:
```diff
@@ -1,2 +1,8 @@
+### git clone https://github.com/aws/aws-lc
+### cd aws-lc
+### git checkout origin/fips-2024-09-27
+### git format-patch --start-number=1001 AWS-LC-FIPS-3.0.0..
```
Please bring back `--no-numbered` to avoid embedding the sequence numbers in the patch files. Otherwise they all get modified whenever a new patch is added.

ytsssun replied:
In that case, do we want to rename the [0001-Avoid-weak-symbols-when-building-for-Go.patch](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/266/files#diff-192be1cf758a4f3f7dddc76949699a4ebf654e92d71f749f2c6e3e77e76701c2) with different prefix? 

Otherwise, there will be two patch files with same 0001 prefix.

ytsssun replied:
Just updated patch with --no-numbered. Full command
```
git format-patch --start-number=1001 --no-numbered
 AWS-LC-FIPS-3.0.0..
```

ytsssun commented:
Tested the change and it worked for me. I will update the patch names per @bcressey
